---
title: "Comparing STAR and Salmon"
output:
    html_document:
      keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```

```{r, include = F}
library(edgeR)
library(ggplot2)
library(gridExtra)
library(dplyr)
```

```{r, include = F}
star.counts <- read.delim("C:/users/bpdurbin/desktop/2019_August_UCD_mRNAseq_Workshop/differential_expression/de_data/rnaseq_workshop_counts.txt", row.names = 1)
salmon.counts <- read.delim("C:/users/bpdurbin/desktop/2019_August_UCD_mRNAseq_Workshop/differential_expression/de_data/salmon_counts.txt", row.names = 1)
snames <- colnames(star.counts) # Sample names
snames
factor1 <- substr(snames, 7, 7) 
factor2 <- substr(snames, 8, 8)
group <- interaction(factor1, factor2)
```

```{r}
d.star <- DGEList(star.counts)
d.star <- calcNormFactors(d.star)
d.salmon <- DGEList(salmon.counts)
d.salmon <- calcNormFactors(d.salmon)
```

### Pairwise correlations of normalized counts on same sample
```{r, include = F}
samples <- colnames(star.counts)[1:4]
genes.in.both <- intersect(rownames(star.counts), rownames(salmon.counts))
cpms.star <- cpm(d.star, log = T)[genes.in.both,]
cpms.salmon <- cpm(d.salmon, log = T)[genes.in.both,]
tmp <- lapply(samples, function(samp){
  df <- data.frame(STAR = cpms.star[,samp], Salmon = cpms.salmon[,samp])
  g <- ggplot(df, aes(x = STAR, y = Salmon)) + geom_point() + ggtitle(samp) + geom_abline(slope = 1, intercept = 0, col = "blue", lty = 2) + xlim(0, 15) + ylim(0, 15) + theme(plot.margin = margin(0,2,2,2))
  return(g)
})
```

```{r, message = F, warning = F}
marrangeGrob(tmp, nrow = 2, ncol = 2)
```

(Figure shows 4 samples, these are representative of pattern on remaining samples.)

Note the pattern of genes with low expression with STAR and high expression with Salmon.

### MDS plot, STAR counts:
```{r}
plotMDS(d.star, col = as.numeric(group))
```

### MDS plot, Salmon counts
```{r}
coords <- plotMDS(d.salmon, col = as.numeric(group))
```

Note that you can rotate or flip an MDS plot without changing the distances between samples (the orientation is arbitrary).  Here's the Salmon MDS plot rotated to match the orientation of the STAR MDS plot:

```{r}
plot(coords$x, -coords$y, xlab = "Leading logFC dim 1", ylab = "Leading logFC dim 2", type = "n")
text(coords$x, -coords$y, labels = colnames(salmon.counts), col = as.numeric(group))
```

```{r, include = F}
DE.star <- read.delim("C:/users/bpdurbin/desktop/2019_August_UCD_mRNAseq_Workshop/differential_expression/de_data/A.C_v_A.D.txt", row.names = 1)
# Get A.C_v_A.D list for salmon
genes.use <- intersect(rownames(DE.star), rownames(salmon.counts))
DE.star <- DE.star[genes.use,]
d.salmon.filtered <- d.salmon[genes.use,]
mm <- model.matrix(~0 + group)
y.salmon <- voom(d.salmon.filtered, mm)
fit <- lmFit(y.salmon, mm)
contr <- makeContrasts(groupA.C - groupA.D, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
DE.salmon <- topTable(tmp, sort.by = "P", n = Inf)

anno <- read.delim("mart_export.txt", stringsAsFactors = F)

DE.star$Gene.stable.ID <- unlist(lapply(strsplit(rownames(DE.star), split = ".", fixed = T), function(x)x[1]))
DE.star <- left_join(DE.star, select(anno, -Gene.stable.ID.version)) %>% select(Gene.stable.ID, HGNC.symbol, everything())

DE.salmon$Gene.stable.ID <- unlist(lapply(strsplit(rownames(DE.salmon), split = ".", fixed = T), function(x)x[1]))
DE.salmon <- left_join(DE.salmon, select(anno, -Gene.stable.ID.version)) %>% select(Gene.stable.ID, HGNC.symbol, everything())
```

### Top 10 genes with STAR
```{r}
head(DE.star, 10)
```

### Top 10 genes with Salmon
```{r}
head(DE.salmon, 10)
```

`r length(which(DE.star$adj.P.Val < 0.05))` genes are DE (adjusted P < 0.05) with STAR, and `r length(which(DE.salmon$adj.P.Val < 0.05))` genes are DE with Salmon.  All genes DE with Salmon are DE with STAR.

### Conclusions
* STAR and Salmon show differences in quantification of low expressed genes.
* This does not seem to have much impact on the relative distances between samples or differential expression (note that low expressed genes are often filtered out before analysis).
* Pick one approach for a project and go with it.
* Don't run multiple analyses and pick the one that gives you the most DE genes.




